internal/timeline/tweener.go

```diff
@@ -0,0 +1,308 @@
+	if start, ok = t.startVal.(float64); !ok {
+		return nil
+	}
+	var end float64
+	if end, ok = t.endVal.(float64); !ok {
+		return nil
+	}
+
+	switch t.mode {
+	case CUR_TO_END, START_TO_CUR:
+		delta := end - start
+		if math.Abs(float64(delta)) < float64(EPSILON) {
+			return nil
+		}
+		if t.duration < EPSILON {
+			return nil
+		}
+		if t.getter == nil || t.setter == nil {
+			return nil
+		}
+		cur, ok2 := t.getter().(float64)
+		if !ok2 {
+			return nil
+		}
+
+		speed := delta / float64(t.duration)
+		need := math.Abs(float64((end - cur) / speed))
+		if *time < TIME(need) {
+			t.setter(float64(float64(cur) + float64(speed)*float64(*time)))
+			*time = 0.0
+			return t
+		} else {
+			*time -= TIME(need)
+			t.setter(t.endVal)
+			return nil
+		}
+	case START_TO_END:
+		delta := end - start
+		if math.Abs(float64(delta)) < float64(EPSILON) {
+			return nil
+		}
+		if t.duration < EPSILON {
+			return nil
+		}
+		if t.setter == nil {
+			return nil
+		}
+		cur, ok2 := t.curVal.(float64)
+		if !ok2 {
+			return nil
+		}
+
+		speed := delta / float64(t.duration)
+		need := math.Abs((end - cur) / speed)
+		if *time < TIME(need) {
+			t.curVal = cur + float64(speed*float64(*time))
+			t.setter(t.curVal)
+			*time = 0.0
+			return t
+		} else {
+			*time -= TIME(need)
+			t.curVal = t.endVal
+			t.setter(t.endVal)
+			return nil
+		}
+	}
+	return nil
+}
+
+func (t *Tweener) onStepV2(time *TIME) ITimeline {
+	var start math32.Vector2
+	var ok bool
+	if start, ok = t.startVal.(math32.Vector2); !ok {
+		return nil
+	}
+	var end math32.Vector2
+	if end, ok = t.endVal.(math32.Vector2); !ok {
+		return nil
+	}
+
+	switch t.mode {
+	case CUR_TO_END, START_TO_CUR:
+		delta := (&end).Sub(&start)
+		if delta.LengthSquared() < float64(EPSILON) {
+			return nil
+		}
+		if t.duration < EPSILON {
+			return nil
+		}
+		if t.getter == nil || t.setter == nil {
+			return nil
+		}
+		cur, ok2 := t.getter().(math32.Vector2)
+		if !ok2 {
+			return nil
+		}
+		speed := delta.Scale(1.0 / float64(t.duration))
+		need := math.Abs((&end).Sub(&cur).Length() / speed.Length())
+		if *time < TIME(need) {
+			v2 := (&cur).Add((speed.Scale(float64(*time))))
+			t.setter(*v2)
+			*time = 0.0
+			return t
+		} else {
+			*time -= TIME(need)
+			t.setter(t.endVal)
+			return nil
+		}
+	case START_TO_END:
+		delta := (&end).Sub(&start)
+		if delta.LengthSquared() < float64(EPSILON) {
+			return nil
+		}
+		if t.duration < EPSILON {
+			return nil
+		}
+		if t.setter == nil {
+			return nil
+		}
+		cur, ok2 := t.curVal.(math32.Vector2)
+		if !ok2 {
+			return nil
+		}
+
+		speed := delta.Scale(1.0 / float64(t.duration))
+		need := math.Abs((&end).Sub(&cur).Length() / speed.Length())
+		if *time < TIME(need) {
+			t.curVal = *((&cur).Add(speed.Scale(float64(*time))))
+			t.setter(t.curVal)
+			*time = 0.0
+			return t
+		} else {
+			*time -= TIME(need)
+			t.curVal = t.endVal
+			t.setter(t.endVal)
+			return nil
+		}
+	}
+	return nil
+}
+
+func (t *Tweener) From() ITweener {
```

Comment:
Is this a clone function?

Reply:
Not a clone function. It's a reverse version of t.